File: packages/prism-core/src/foundation/date/date-format.ts
// @prism/core/date — Date component extraction and display utilities.

import { parseDate, formatDate, addDays, daysInMonth } from "./date-math.js";

/** Day of week for a YYYY-MM-DD string: 0=Sunday … 6=Saturday (UTC). */
export function dayOfWeek(iso: string): number {
  return new Date(parseDate(iso)).getUTCDay();
}

/** Return the Monday (start of ISO week) for the given YYYY-MM-DD. */
export function weekStart(iso: string): string {
  const dow = dayOfWeek(iso);
  const offset = dow === 0 ? 6 : dow - 1;
  return addDays(iso, -offset);
}

/** Return the Sunday (end of ISO week) for the given YYYY-MM-DD. */
export function weekEnd(iso: string): string {
  const start = weekStart(iso);
  return addDays(start, 6);
}

/** Return the first day of the month for the given YYYY-MM-DD. */
export function monthStart(iso: string): string {
  const [year = 0, month = 1] = iso.split("-").map(Number);
  return formatDate(Date.UTC(year, month - 1, 1));
}

/** Return the last day of the month for the given YYYY-MM-DD. */
export function monthEnd(iso: string): string {
  const [year = 0, month = 1] = iso.split("-").map(Number);
  const lastDay = daysInMonth(year, month - 1);
  return formatDate(Date.UTC(year, month - 1, lastDay));
}

/** Return the first day of the quarter for the given YYYY-MM-DD. */
export function quarterStart(iso: string): string {
  const [year = 0, month = 1] = iso.split("-").map(Number);
  const quarterFirstMonth = Math.floor((month - 1) / 3) * 3 + 1;
  return formatDate(Date.UTC(year, quarterFirstMonth - 1, 1));
}

/** Return the first day of the year for the given YYYY-MM-DD. */
export function yearStart(iso: string): string {
  const year = Number(iso.split("-")[0]);
  return formatDate(Date.UTC(year, 0, 1));
}

/**
 * Format a YYYY-MM-DD string for display (e.g. "Mar 15, 2026").
 * Uses Intl.DateTimeFormat — locale-aware but UTC-pinned.
 */
export function formatDisplayDate(iso: string, locale?: string): string {
  const ms = parseDate(iso);
  return new Intl.DateTimeFormat(locale ?? "en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
    timeZone: "UTC",
  }).format(ms);
}

/** Format a YYYY-MM-DD string as a short display (e.g. "Mar 15"). */
export function formatShortDate(iso: string, locale?: string): string {
  const ms = parseDate(iso);
  return new Intl.DateTimeFormat(locale ?? "en-US", {
    month: "short",
    day: "numeric",
    timeZone: "UTC",
  }).format(ms);
}

/** Return the ISO year (YYYY) as a number. */
export function getYear(iso: string): number {
  return Number(iso.split("-")[0]);
}

/** Return the ISO month (1–12). */
export function getMonth(iso: string): number {
  return Number(iso.split("-")[1]);
}

/** Return the ISO day of month (1–31). */
export function getDay(iso: string): number {
  return Number(iso.split("-")[2]);
}

File: packages/prism-core/src/foundation/date/date_format.rs
//! Date component extraction and display utilities. Port of
//! `foundation/date/date-format.ts`.
//!
//! The legacy TS module relies on `Intl.DateTimeFormat` for the
//! localized display helpers; the Rust port hardcodes the
//! equivalent `en-US` short-month output to avoid pulling in a
//! locale crate. When we add `icu` later the locale argument can
//! become meaningful.

use chrono::{Datelike, Duration, NaiveDate, Weekday};

use super::date_math::{add_days, days_in_month, format_date, parse_date};

/// Day of week for an ISO date: `0` = Sunday, `6` = Saturday
/// (matching JS `Date.getUTCDay()`).
pub fn day_of_week(iso: &str) -> u32 {
    match parse_date(iso).weekday() {
        Weekday::Sun => 0,
        Weekday::Mon => 1,
        Weekday::Tue => 2,
        Weekday::Wed => 3,
        Weekday::Thu => 4,
        Weekday::Fri => 5,
        Weekday::Sat => 6,
    }
}

/// Monday (ISO week start) for the given date.
pub fn week_start(iso: &str) -> String {
    let dow = day_of_week(iso);
    let offset = if dow == 0 { 6 } else { dow - 1 };
    add_days(iso, -(offset as i64))
}

/// Sunday (ISO week end) for the given date.
pub fn week_end(iso: &str) -> String {
    add_days(&week_start(iso), 6)
}

pub fn month_start(iso: &str) -> String {
    let d = parse_date(iso);
    format_date(NaiveDate::from_ymd_opt(d.year(), d.month(), 1).unwrap_or_default())
}

pub fn month_end(iso: &str) -> String {
    let d = parse_date(iso);
    let last_day = days_in_month(d.year(), d.month() - 1);
    format_date(NaiveDate::from_ymd_opt(d.year(), d.month(), last_day).unwrap_or_default())
}

pub fn quarter_start(iso: &str) -> String {
    let d = parse_date(iso);
    let quarter_first_month = ((d.month() - 1) / 3) * 3 + 1;
    format_date(NaiveDate::from_ymd_opt(d.year(), quarter_first_month, 1).unwrap_or_default())
}

pub fn year_start(iso: &str) -> String {
    let d = parse_date(iso);
    format_date(NaiveDate::from_ymd_opt(d.year(), 1, 1).unwrap_or_default())
}

/// Long display form ("Mar 15, 2026"). The `_locale` argument is
/// accepted for API parity with the legacy TS signature but is
/// currently ignored — the output is always en-US style.
pub fn format_display_date(iso: &str, _locale: Option<&str>) -> String {
    let d = parse_date(iso);
    format!("{} {}, {}", short_month(d.month()), d.day(), d.year())
}

pub fn format_short_date(iso: &str, _locale: Option<&str>) -> String {
    let d = parse_date(iso);
    format!("{} {}", short_month(d.month()), d.day())
}

pub fn get_year(iso: &str) -> i32 {
    parse_date(iso).year()
}

pub fn get_month(iso: &str) -> u32 {
    parse_date(iso).month()
}

pub fn get_day(iso: &str) -> u32 {
    parse_date(iso).day()
}

fn short_month(month: u32) -> &'static str {
    match month {
        1 => "Jan",
        2 => "Feb",
        3 => "Mar",
        4 => "Apr",
        5 => "May",
        6 => "Jun",
        7 => "Jul",
        8 => "Aug",
        9 => "Sep",
        10 => "Oct",
        11 => "Nov",
        12 => "Dec",
        _ => "",
    }
}

// `Duration` is imported for the doc rendering of `add_days`
// callers; suppress the unused-import warning with a dummy ref.
const _: fn() -> Duration = || Duration::days(0);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn week_start_returns_monday() {
        // 2026-03-15 is a Sunday.
        assert_eq!(week_start("2026-03-15"), "2026-03-09");
    }

    #[test]
    fn month_end_handles_leap_february() {
        assert_eq!(month_end("2024-02-10"), "2024-02-29");
        assert_eq!(month_end("2025-02-10"), "2025-02-28");
    }

    #[test]
    fn quarter_start_snaps_to_quarter() {
        assert_eq!(quarter_start("2026-01-15"), "2026-01-01");
        assert_eq!(quarter_start("2026-05-15"), "2026-04-01");
        assert_eq!(quarter_start("2026-08-15"), "2026-07-01");
        assert_eq!(quarter_start("2026-11-15"), "2026-10-01");
    }

    #[test]
    fn display_formats_match_legacy_en_us() {
        assert_eq!(format_display_date("2026-03-15", None), "Mar 15, 2026");
        assert_eq!(format_short_date("2026-03-15", None), "Mar 15");
    }
}
